### ISSUE TITLE
Need general way to configure command editors

### ISSUE THREAD

**Issue by smalers** [OpenWaterFoundation/owf-app-geoprocessor-python#152]
The GeoProcessor design uses general command editors so that custom editors don't need to be developed for each command.  This reduces custom coding for editors but poses some design challenges for how to generically specify properties to result in custom layouts, etc.

One approach that could be worthwhile is to use a JSON-like hierarchy of properties.  For example, define a dictionary in the command class with information that will display in the editor.  This is OK to achieve separating UI from command class because the data will consist of primitives like strings.  For example, define a dictionary that contains a hierarchy of property strings:

command_metadata = {}
command_metadata['description'] = 'Short description used at top of editor'
command_metadata['editortype'] = 'simple'

parameter_metadata = {}
parameter_metadata['ParameterName.group'] = 'The group/tab in which a parameter is grouped'
parameter_metadata['ParameterName.description'] = 'short description used to right of UI component'
parameter_metadata['ParameterName.tooltip'] = 'tooltip for UI'
parameter_metadata['ParameterName.default'] = 'Default value for parameter'

The following are design decisions to address:

* How to easily extract the parameter names for a command (is there a property that is always present, such as `required`)?
* Do text values optionally contain HTML or other formatting (is this supported in QT5)?
* How about embedded URLs, such as links to third-party documentation?  One way to address is to include such things in the MkDocs documentation that is accessible from the ***Documentation*** button.
* Would need to add utility functions to handle the lookups, for example in `command_util`.
* How to handle choices?
* How to handle more complex logic like parameter values that are determined at runtime, for example list of available GeoLayer ID?
* Qt5 may already have some built-in tools for such things, such as UI model definitions in XML.  Should we use such things or code ourselves?
* Worst case, if an editor is complicated, then build a custom editor.  Ideally, the generic editor UI can work for most commands.

**Comment by smalers** [comment]
Below are notes on possible properties.

# Properties for Edtior Dialogs #

These properties can be specified in the command class for the command and properties
in order to allow the editor classes to layout and provide appropriate input input components.
The data can be maintained in the `AbstractCommand`.

```
# Dictionary of string properties to describe command metadata.
command_metadata = {}

# Dictionary of string properties to describe command parameters.
parameter_metadata = {}
```

| **Command Property** | **Description** | **Default** |
|----------------------|-----------------|-------------|
| `Description`        | Description of the command to display at the top of the command dialog. Can this be HTML with embedded link? | Command name. |
| `EditorType`         | What type of editor to use, `Simple` or `Tabbed` (maybe make this an enumeration). |

| **Parameter Property** | **Description** | **Default** |
|------------------------|-----------------|-------------|
| `ParameterName.Group`  | The group (tab) used to group parameters, such as `Input`. | No group used - use simple editor. |
| `ParameterName.Description` | The description to be shown on the right-side of the editor. | No description is shown. |
| `ParameterName.Label` | The label to be shown to the left` of the component. | Parameter name. |
| `ParameterName.Tooltip` | The tooltip to be shown when moused over the input component. | No tooltip is shown. |
| `ParameterName.Required` | Indicate whether required:  `Required`, `Optional`, or maybe something more complex. | Must be specified. |
| `ParameterName.Values` | Indicate a list of values to show in choices, may need a way to dynamically populate (lambda?). |
| `ParameterName.DefaultValue` | The default value as a string, as if the user entered into the component, to be shown in the description. |
| `ParameterName.FileSelectorType` | Indicate the type of selector as enumeration `Read`, `Write`.  Will result in `...` button to browse for fall. |
| `ParameterName.?` | Need some way to indicate default file extension, recognized extensions and descriptions | |
| `ParameterName.?` | Perhaps need a way to set the dimention of a text field or text area. Dialogs are kind of ugly when text fields span the entire width.  Maybe use the concept of number of columns? | |

The parameter name needs to be part of the property to ensure that the relationships are describe in flat form.
The `command_util` module can be updated with methods to extract from the metadata the dictionary value,
for example to get the commands in the proper order.

The format of the description should be `Required - Description (default=DefaultValue)`,
with intelligence to handle None and empty strings.

The validation for the component should be able to occur using exising command editor features.
This can validate whether input is a number, etc.
It may be possible to use the metadata to automatically implement validators but that does not need to happen now.

**Comment by smalers** [comment]
Here is my feedback on the WebGet command.

1. I tend to use some type of punctuation at the end of text to show that the text is done.  This helps users know that nothing more comes after.   Therefore:
    1. Make sure the description has a period at the end.
    2. Make sure that descriptions have a period at the end.
    3. Make sure tooltip text has period at the end.
2. Need to adopt a standard for linebreaks such as the command description.  I recommend embedding '\n'.
3. Why does the font seem small for dialog labels and field text.  The UI menu and command list text size seems comfortable but some of the other text seems small.  Maybe add a separate issue to evaluate this.  Is there a style file somewhere that controls global styles?  This could be an issue on high-resolution displays where text can be small.
4. The label on the left side of a component should be allowed to be different than the parameter name and be more user-friendly.  We don't want to force parameter names on users everywhere.  For example, the label should be "Output File:" or maybe "Output file:".  I researched standards for TSTool and settled on a style with label uppercase.
5. See issue #xxx about setting the editor titles.  Do this now so it is consistent.
6. UserName does not show a tooltip.
7. Note that some comboboxes are editable and some are not.  This should be a parameter that is configurable.  An editable combobox recognizes that some standard values are selectable, but the user can override with their own value.
8. The indication of case-sensitive for User name and password might be better in a tooltip.
9. As discussed, need to be thoughtful about how default value is shown and used.  Maybe use something like `DefaultValue` for the internal default value and `DefaultDisplayValue` for what the user sees.  For example, they might see `None` but internally that translates to a blank string.
10. Because filenames can be long, the approach taken in TSTool is to use some of the description space on the right to have more layout columns.  For example, the description in the layout might allow for 2 or more columns of layout.  Then the filename could take the left-most columns and the `...` could take the right-most single column in the layout.  The tooltip can be used to explain the filename.  We could choose to at least show `Optional` or `Required` on the far right.  The beauty of handling the layout in one place is we can evolve this in one place.  One comment is that for layout purposes, more than one column can be allocated even if not needed now.  This gives flexibility to split later for other UI components that need more columns.
11. Try to minimize extra buffer space.  This is more important for busy editors that have a lot of input components.
12.  The tooltip on the `...` should be more specific such as `Select URL to download from.`  The dialog does not actually open the file, it helps select a filename.
13.  The URL should not be a browser right since you can't browse the web easily.
14.  Resize seems OK other than command text area needs to expand to fill the space.  The Python code behave better than Java - in Java I often disabled resize because it behaved badly.
15.  For filenames, default to relative path relative to the command file, which will be set as the application working directory.  See how I do in TSTool.

**Comment by smalers** [comment]
Here is my feedback from a review of the current build:

First I loaded the test command file `Webget/test-WebGet-DownloadableFile.gp`:

1. Clicking on ***Documentation*** crashes the application.  This also occurred on other commands.
2. I edited comments and added comment lines, etc.  The functionality seems to work but see repeated messages:  `INFO:geoprocessor.ui.core.GeoProcessorCommandEditorFactory:Command line is unknown syntax. Creating GenericCommandEditor for: # Remove the downloadable file`.   Make sure that the comments are being handled properly including with indentation.
